### DISCUSSION TITLE
DynamoDB outbox implementation issue

### DISCUSSION THREAD

**Discussion by romtur** [BrighterCommand/Brighter#3091] [Q&A]
We've been using Brighter along with Kafka for some time now. A week ago, we updated Brighter to version 1.0.0-preview.3.
Recently, we decided to implement the outbox. The requirement is to go with a NoSQL approach, so we chose DynamoDB, even though we are using MySQL as our entity store.

I was following this sample: https://github.com/BrighterCommand/Brighter/tree/master/samples/WebAPI_Dynamo
And this is my outbox setup:

```  
brighterBuilder.UseExternalBus((configure) =>
      { 
          configure.ProducerRegistry = new KafkaProducerRegistryFactory(kafkaConfiguration, publications).Create();
          configure.Outbox = new DynamoDbOutbox(dynamoDBClient, new DynamoDbConfiguration());
          configure.TransactionProvider = typeof(DynamoDbUnitOfWork);
          configure.ConnectionProvider = typeof(DynamoDbUnitOfWork);
      });
```

When I'm trying to make PostAsync or DepositPostAsync (tried both) I'm getting this exception:

`System.InvalidCastException: Unable to cast object of type 'Paramore.Brighter.ExternalBusServices2[Paramore.Brighter.Message,Amazon.DynamoDBv2.Model.TransactWriteItemsRequest]' to type 'Paramore.Brighter.ExternalBusServices2[Paramore.Brighter.Message,System.Transactions.CommittableTransaction]'`

I assume the reason could be that I cannot use the **DynamoDbUnitOfWork** as a **TransactionProvider** because there are no live DynamoDB transactions. However, when I comment out this line, **_asyncOutbox** is always null in the **DepositPostAsync** method, resulting in nothing being written to the Outbox DB.

I would greatly appreciate any advice or help.
What I aim to achieve is to properly implement the outbox using the **Deposit and Clear** approach. Is it possible to do this with MySQL as the entity store and DynamoDB as the outbox?

**Comment by iancooper** [comment]
HI @romtur I will check. As this is a preview release, there could be an issue. One thing to make sure of, is that when you have configured the ExternalBus, you have configured the ConnectionProvider and TransactionProvider as per https://github.com/BrighterCommand/Brighter/blob/a74571c063f892320dbe6cf128f04ddc619e8c8b/samples/WebAPI_Dynamo/GreetingsWeb/Startup.cs#L185

**Comment by romtur** [comment]
Hi @iancooper 

It looks like it's an issue of version 10
Only the **PostAsync** (**Post** approach when there is no transaction)

`System.InvalidCastException: Unable to cast object of type 'Paramore.Brighter.ExternalBusServices2[Paramore.Brighter.Message,Amazon.DynamoDBv2.Model.TransactWriteItemsRequest]' to type 'Paramore.Brighter.ExternalBusServices2[Paramore.Brighter.Message,System.Transactions.CommittableTransaction]'`

This failure occurs in **CommandProcessor**:

![image](https://github.com/BrighterCommand/Brighter/assets/22519624/686890e6-fc48-4f2e-8081-91b589a4d951)

I tried to call DepositPostAsync (**Deposit and Clear** approach), providing **IAmADynamoDbTransactionProvider** and it works fine

Also tried the same **PostAsync** on version 9.8.0 and it works fine as well.

Should we create an issue for this?

Best Regards,
Roman Turchyniak

**Reply by iancooper** [reply]
Yes please do. We have changed some of this for V10 to make the interface simpler, so we probably broke something and I'll need to have a look.
